File: src/nn/modules/normalization/GroupNorm.cpp
#include "nn/modules/normalization/GroupNorm.hpp"
#include "nn/Parameter.hpp"
#include "ggml/Context.hpp"

GroupNorm::GroupNorm(int64_t num_groups, int64_t num_channels, float eps, bool affine, bool bias)
    : num_groups_(num_groups),
      eps_(eps),
      affine_(affine),
      bias_(bias)
{
    if (affine_) {
        modules["weight"] =
            std::make_shared<Parameter>(
                Tensor::Shape({num_channels})
            );

        if (bias_)
            modules["bias"] =
                std::make_shared<Parameter>(
                    Tensor::Shape({num_channels})
                );
    }
}

Tensor GroupNorm::forward(
    Context& context,
    Tensor input
) {
    auto shape = input.shape();

    const int64_t batch = shape[0];
    const int64_t channels = shape[1];
    const int64_t height = shape[2];
    const int64_t width = shape[3];


    /*
        PyTorch:

        N,C,H,W
          |
          v
        N,G,C/G*H*W

        Let reshape infer the last dimension.
    */
    input = input.reshape({
        batch,
        num_groups_,
        -1
    });


    input = Tensor(
        *context,
        ggml_norm(
            *context,
            *input,
            eps_
        ),
        input.shape()
    );


    input = input.reshape({
        batch,
        channels,
        height,
        width
    });


    if (affine_) {

        auto weight =
            std::static_pointer_cast<Parameter>(
                modules["weight"])
            ->forward();


        weight = weight.reshape({
            1,
            channels,
            1,
            1
        });


        input = input * weight;


        if (bias_) {
            auto bias =
                std::static_pointer_cast<Parameter>(
                    modules["bias"])
                ->forward();


            bias = bias.reshape({
                1,
                channels,
                1,
                1
            });


            input = input + bias;
        }
    }


    return input;
}

File: test/src/nn/modules/normalization/GroupNorm.py
from utils import TestCase
import torch
from torch.nn.modules.normalization import GroupNorm

class TestNNNormalizationGroupNorm(TestCase):
    def test_default(self):
        model = GroupNorm(
            num_groups=2,
            num_channels=4,
            eps=1e-6,
        )

        input = torch.randn(2, 4, 3, 3)

        expected = model.forward(input)

        actual = self.cli(
            "GroupNorm",
            "--num_groups", "2",
            "--num_channels", "4",
            "--eps", "1e-6",
            "--input", str(input.tolist()),
            "--param-weight", str(model.weight.tolist()),
            "--param-bias", str(model.bias.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_multiple_groups(self):
        model = GroupNorm(
            num_groups=4,
            num_channels=8,
            eps=1e-6,
        )

        input = torch.randn(1, 8, 4, 4)

        expected = model.forward(input)

        actual = self.cli(
            "GroupNorm",
            "--num_groups", "4",
            "--num_channels", "8",
            "--eps", "1e-6",
            "--input", str(input.tolist()),
            "--param-weight", str(model.weight.tolist()),
            "--param-bias", str(model.bias.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_no_affine(self):
        model = GroupNorm(
            num_groups=2,
            num_channels=4,
            eps=1e-6,
            affine=False,
        )

        input = torch.randn(2, 4, 2, 2)

        expected = model.forward(input)

        actual = self.cli(
            "GroupNorm",
            "--num_groups", "2",
            "--num_channels", "4",
            "--eps", "1e-6",
            "--affine", "false",
            "--input", str(input.tolist()),
        )

        self.assertTensors(actual, [expected])

    def test_no_bias(self):
        model = GroupNorm(
            num_groups=2,
            num_channels=4,
            eps=1e-6,
            bias=False,
        )

        input = torch.randn(2, 4, 2, 2)

        expected = model.forward(input)

        actual = self.cli(
            "GroupNorm",
            "--num_groups", "2",
            "--num_channels", "4",
            "--eps", "1e-6",
            "--bias", "false",
            "--input", str(input.tolist()),
            "--param-weight", str(model.weight.tolist()),
        )

        self.assertTensors(actual, [expected])
